<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Normalise and subset the data in <code>mmt</code> objects based on variable/sample metadata. — mt_subset • mmtravis</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Normalise and subset the data in <code>mmt</code> objects based on variable/sample metadata. — mt_subset" />

<meta property="og:description" content="Subset mt objects based on sample metadata" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmtravis</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Normalise and subset the data in <code>mmt</code> objects based on variable/sample metadata.</h1>
    
    <div class="hidden name"><code>mt_subset.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Subset mt objects based on sample metadata</p>
    
    </div>

    <pre class="usage"><span class='fu'>mt_subset</span>(<span class='no'>data</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mmt</th>
      <td><p>(<em>required</em>) Data list as loaded with <code><a href='mt_load.html'>mt_load</a></code>.</p></td>
    </tr>
    <tr>
      <th>sub_genes</th>
      <td><p>(<em>optional</em>) A string specifying a logical row subset operation on the mtgene dataframe in the mt object parsed to <a href='http://www.rdocumentation.org/packages/base/topics/subset'>subset</a>.</p></td>
    </tr>
    <tr>
      <th>sub_samples</th>
      <td><p>(<em>optional</em>) A string specifying a logical row subset operation on the mtmeta dataframe in the mt object parsed to <a href='http://www.rdocumentation.org/packages/base/topics/subset'>subset</a>.</p></td>
    </tr>
    <tr>
      <th>minreads</th>
      <td><p>Minimum average number of reads pr. gene. Genes below this value will be removed. (<em>default:</em> <code>0</code>)</p></td>
    </tr>
    <tr>
      <th>frac0</th>
      <td><p>Fraction of zeros allowed per gene. Genes with a higher fraction of zeros will be removed. (<em>default:</em> <code>1</code>)</p></td>
    </tr>
    <tr>
      <th>normalise</th>
      <td><p>Normalise the read counts AFTER reads have been removed by the minreads argument but BEFORE any sample/gene subsetting. (<em>default:</em> <code>"none"</code>)</p><ul>
<li><p><code>"quantile"</code>: Quantile normalization. See <a href='http://www.rdocumentation.org/packages/preprocessCore/topics/normalize.quantiles'>normalize.quantiles</a> for details.</p></li>
<li><p><code>"TPM"</code>: Normalise read counts to Transcripts Per Milion (TPM).</p></li>
<li><p><code>"abundance"</code>: Normalise read counts to relative abundance.</p></li>
<li><p><code>"libsize"</code>: Normalise the read counts to adjust for gene dispersion and total read counts per sample. See <a href='http://www.rdocumentation.org/packages/DESeq2/topics/estimateSizeFactorsForMatrix'>estimateSizeFactorsForMatrix</a> for details.</p></li>
<li><p><code>"vst"</code>: Normalise as "libsize" and perform robust log2-transformation. See <a href='http://www.rdocumentation.org/packages/DESeq2/topics/vst'>vst</a> for details.</p></li>
<li><p><code>"log2"</code>: Normalise as "libsize" and perform log2(x + 1)-transformation.</p></li>
<li><p><code>"none"</code>: No normalisation.</p></li>
</ul></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A modifed mt object</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function <code>mt_subset</code> operates in three steps:</p><ul>
<li><p><strong>Filter</strong> - genes according to <code>minreads</code> and <code>frac0</code>.</p></li>
<li><p><strong>Normalise</strong> - as specified in <code>normalise</code>.</p></li>
<li><p><strong>Subset</strong> - according to <code>sub_genes</code> and <code>sub_samples</code>.</p></li>
</ul>
    <p>Subsetting are performed on the mtmeta/mtgene data by <a href='http://www.rdocumentation.org/packages/base/topics/subset'>subset</a> and the whole object is then adjusted accordingly.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
</div><span class='co'># NOT RUN {</span>
<span class='co'># Get some data.</span>
<span class='fu'>data</span>(<span class='st'>"example_mmt"</span>)

<span class='co'># Let's subset to contig 1, 7675, and 69676.</span>
<span class='no'>mt1</span> <span class='kw'>&lt;-</span> <span class='fu'>mt_subset</span>(<span class='no'>example_mmt</span>,<span class='kw'>sub_genes</span> <span class='kw'>=</span> <span class='st'>"contig %in% c(1,7675,69676)"</span>)
<span class='no'>mt1</span>

<span class='co'># Let's subset to specific organism.</span>
<span class='no'>mt2</span> <span class='kw'>&lt;-</span> <span class='fu'>mt_subset</span>(<span class='no'>example_mmt</span>,<span class='kw'>sub_samples</span> <span class='kw'>=</span> <span class='st'>"Organism == 'Brocadia'"</span>)
<span class='no'>mt2</span>

<span class='co'># Let's do both and remove genes with less than 10000 reads in total.</span>
<span class='no'>mt3</span> <span class='kw'>&lt;-</span> <span class='fu'>mt_subset</span>(<span class='no'>example_mmt</span>,
  <span class='kw'>sub_samples</span> <span class='kw'>=</span> <span class='st'>"Organism == 'Brocadia'"</span>,
  <span class='kw'>sub_genes</span>   <span class='kw'>=</span> <span class='st'>"contig %in% c(1,7675,69676)"</span>,
  <span class='kw'>minreads</span>    <span class='kw'>=</span> <span class='fl'>10000</span>)
<span class='no'>mt3</span>

<span class='co'># You can also normalise the data and subset.</span>
<span class='no'>mt4</span> <span class='kw'>&lt;-</span> <span class='fu'>mt_subset</span>(<span class='no'>example_mmt</span>,
  <span class='kw'>sub_samples</span> <span class='kw'>=</span> <span class='st'>"Organism == 'Brocadia'"</span>,
  <span class='kw'>sub_genes</span>   <span class='kw'>=</span> <span class='st'>"contig %in% c(1,7675,69676)"</span>,
  <span class='kw'>minreads</span>    <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>normalise</span>   <span class='kw'>=</span> <span class='st'>"libsize"</span>)
<span class='no'>mt4</span> <span class='co'># Note "Normalised:" is now included.</span>
<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Thomas Yssing Michaelsen <a href='mailto:tym@bio.aau.dk'>tym@bio.aau.dk</a>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Thomas Yssing Michaelsen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

